### PR TITLE
Add PerlIO::via to File::BOM preload

### PR DESCRIPTION
**Pull request by shawnlaffan**
File::BOM can be used as a layer in open(), e.g. ```open my $fh, '<via(File::BOM)', $somefile;```

However, PerlIO::via is not used in the File::BOM code, so Module::ScanDeps has no way of knowing about it unless the script it is executed and attempts to open a file.  A preload allows support for static scanning.  

Let me know if you want a test for this.  I think none of the preloads have tests, but it perhaps would still be useful.

### CODE REVIEW

**Comment by rschupp** [comment]
Hi Shawn, what is the bug this is trying to solve?
If I run `scandeps.pl -B` on 
```
use strict;
use warnings;
use File::BOM;
open my $fh, "<:via(File::BOM)", "bommel.txt";
print <$fh>;
close $fh;
```
it lists
```
'File::BOM'          => '0.15',
'PerlIO'             => '1.10',
'PerlIO::scalar'     => '0.26',
'PerlIO::via'        => '0.16',
```
Or am I missing something here?

**Comment by shawnlaffan** [comment]
In part it seems to be trying to solve my errors in reproduction.  The one-liner I tested with was missing too many of the sanity checks.

That said, the origin of the issue is a mac binary built using pp.  That binary is missing PerlIO::via, so files cannot be opened.  The same code base packs without issues on windows (same version of Module::ScanDeps).  I don't have a mac handy right now, but will double check when I'm back in the office in a few days.

**Comment by shawnlaffan** [comment]
TL;DR - After further checking, it is due to rlib being used to add paths to ```@INC```, which scandeps does not handle.  I'll close this PR since it does not actually solve the root cause,  but maybe there is something that could be added to do so. 

The issue is more generic than just macs.  I added a workaround to call open with the '<:via(File::BOM)' layer some years ago to my PC and Linux build script, but the mac build is more recent - it uses its own script and missed the workaround.

To reproduce the issue, generate a file structure with bin\fb.pl and lib\UseFileBOM.pm (contents below).

If I run scandeps with the -x flag then it finds File::BOM since it is added to %INC when UseFileBOM is loaded.  However, scandeps does not search the source code since it did not find it.  The problem then is that the '<:via(File::BOM)' call is not explicitly exercised during the scan process, so PerlIO and PerlIO::via are not added to %INC.  

If I add the ./lib dir to the PERL5LIB env var then scandeps searches UseFileBOM.pm.

If I turn on the verbose flag then scandeps warns about not finding UseFileBOM.pm, so there was some feedback all along.

I don't know the internals very well, but one possible solution is to double check for any entries in %INC that have not been searched when the -x flag is specified, e.g. because they have been dynamically added during execution.  If these are scanned then more dependencies will be picked up.  

There also seems to be code to handle the ```use lib '../lib'``` case, so perhaps that could be extended to also handle rlib.


Files:

bin/fb.pl
```
use strict;
use warnings;

use rlib;
use UseFileBOM;

if ($ARGV[0]) {
    my $line = UseFileBOM::open_file($0);
}
```

lib/UseFileBOM.pm:
```
package UseFileBOM;

use strict;
use warnings;

use File::BOM qw / :subs /;

sub open_file {
    my $file = shift;
    open my $fh, '<:via(File::BOM)', $file or die $!;
    my $line = <$fh>;
    close $fh;
    return $line;
}

1;

```

**Comment by rschupp** [comment]
> There also seems to be code to handle the use lib '../lib' case, so perhaps that could be extended to also handle rlib.

Well, looking at rlib.pm, I guesstimate that the code to "understand" lines like
```
use rlib;
use rlib "mydir";
```
would be the same size as rlib.pm itself plus I would have to invent stuff to track the current file
into ```scan_chunk()```. Not really worth it.

**Comment by shawnlaffan** [comment]
scan_line seems to only be called from scan_chunk, so it should be possible to also pass the file name as an argument.  Duplicating all of rlib.pm would not be future proof, though, but it has not changed in 20 years so perhaps the risk is low.  If I get a chance I'll try to work up a proof of concept.  

As an alternative, it might be worth warning more loudly if a called package cannot be found.  At the moment it is only done under ```scandeps -v```, and pp seems not to warn at all.

**Comment by rschupp** [comment]
> scan_line seems to only be called from scan_chunk, so it should be possible to also pass the file name as an argument.

```scan_line``` is part of the public API, so that's out of the question.

**Comment by shawnlaffan** [comment]
It would be possible if ```scan_line``` it were converted into a wrapper sub, with the logic in a new sub that handles two arguments.  That still leaves the replication of rlib issue, though.  I have a workaround for my particular use-case, so I think now that I'll leave it be.

WRT the louder warnings on not finding libs, I'll raise the topic on the mailing list or as an issue if I get a chance to look into it.  It's a different topic than this PR (which is closed anyway).
